aead/src/lib.rs

```diff
@@ -0,0 +1,214 @@
+    )]
+    PlaintextTooLarge,
+
+    /// The MAC failed to validate. This should be used to indicated a modified
+    /// ciphertext, modified additional data, or key mismatch.
+    #[fail(display = "The MAC failed to validate")]
+    MacFailure,
+    /// There was an error parsing the padding of this message
+    #[fail(display = "The padding was not correct")]
+    PaddingFailure,
+}
+
+/// An enum describing possible failure modes
+#[cfg(not(feature = "serde"))]
+#[derive(Clone, Copy, Debug, Eq, Fail, Hash, Ord, PartialEq, PartialOrd)]
+pub enum AeadError {
+    /// The nonce provided is the wrong size for this algorithm.
+    #[fail(display = "The nonce given is the wrong size for the algorithm")]
+    InvalidNonceSize,
+    /// The ciphertext buffer is not a multiple of the block size
+    #[fail(display = "The buffer given is the wrong size for the algorithm")]
+    InvalidBufferSize,
+    /// The "bytes used" provided is larger than the provided buffer
+    #[fail(display = "The buffer usage is larger than the buffer size")]
+    InvalidBufferUsed,
+    /// The authentication tag is not the correct size
+    #[fail(display = "The authentication tag is the wrong size for the algorithm")]
+    InvalidTagSize,
+    /// The cipher key was not the size of a block
+    #[fail(display = "The cipher key given on creation was the wrong size")]
+    InvalidCipherKeySize,
+    /// The initialization vector was not the size of a block
+    #[fail(display = "The IV given on creation was the wrong size")]
+    InvalidIvSize,
+    /// The authentication tag is not the size expected by this algorithm
+    #[fail(display = "The key used for the authentication algorithm was the wrong size")]
+    InvalidTagKeySize,
+
+    #[fail(
+        display = "The given ciphertext is larger than the maximum size allowed by this algorithm"
+    )]
+    CiphertextTooLarge,
+    #[fail(
+        display = "The given plaintext is larger than the maximum size allowed by this algorithm"
+    )]
+    PlaintextTooLarge,
+
+    /// The MAC failed to validate. This should be used to indicated a modified
+    /// ciphertext, modified additional data, or key mismatch.
+    #[fail(display = "The MAC failed to validate")]
+    MacFailure,
+    /// There was an error parsing the padding of this message
+    #[fail(display = "The padding was not correct")]
+    PaddingFailure,
+}
+
+/// A trait used to tag ciphertext inputs/outputs for an authenticated block cipher
+pub trait Ciphertext:
+    Clone
+    + Debug
+    + Display
+    + Into<Vec<u8>>
+    + Send
+    + Sized
+    + Sync
+    + for<'bytes> TryFrom<&'bytes [u8], Error = AeadError>
+{
+    /// Retrieve the length of the ciphertext, in bytes
+    fn len(&self) -> usize;
+}
+
+/// An explicit cryptographic nonce, which may be optionally used for
+/// authenticated block ciphers.
+pub trait Nonce:
+    Clone
+    + Debug
+    + Display
+    + Into<Vec<u8>>
+    + Send
+    + Sized
```

Comment:
This bound disallows the use of variable-width nonces. Whether these are a good idea is debatable, but I'll point out this trait is incompatible with Miscreant in its current form:

https://github.com/miscreant/miscreant.rs/blob/master/src/aead.rs#L57